Make reviewer pending rejection input a datetime widget and allow changing it through an action

diox opened the pull request:
Fixes https://github.com/mozilla/addons/issues/14920

### Description

This allows reviewers to change the pending rejection date of versions pending rejection through a widget where they can select the date and time.

Because the actions menu is already pretty large, the clear pending rejection date action is reused, with an option to change or clear ; to tie everything together and make the UI more consistent, the rejection action use the same widgets, making the date choice absolute rather than relative.

### Screenshots

#### New reject multiple versions UI
![Screenshot 2025-01-03 at 17-17-44 Ui-Addon-Install – Add-ons for Firefox](https://github.com/user-attachments/assets/3539f065-b883-4605-aea1-b23d4bb2a1b5)

#### Modified Clear pending rejection action (now Change pending rejection) UI
![Screenshot 2025-01-03 at 17-18-37 Ui-Addon-Install – Add-ons for Firefox](https://github.com/user-attachments/assets/7348ea6f-d9ab-485f-94e0-1daf660c382d)

## Testing

See tests. Non-exhaustive list of scenarios, after submitting an add-on and going to its review page:
- Use the reject multiple versions action with immediate effect
- Use the reject multiple versions action with a delay, and then...
  - Use the change pending rejection action to clear the delay
  - Use the change pending rejection action to change the delay to a date in the future

Changing the delay (not clearing it) should send an email to the developer following copy that can be found as a comment in the issue.

Changing a delay should require the new date to be at least one extra day in the future compared to now. Changing the delay of multiple versions at once is possible, but the existing pending rejection date should all be the same.

diox commented on src/olympia/reviewers/views.py:
```diff
@@ -1114,21 +1114,6 @@ def allow_resubmission(self, request, **kwargs):
-        methods=['post'],
-        permission_classes=[GroupPermission(amo.permissions.REVIEWS_ADMIN)],
-    )
-    def clear_pending_rejections(self, request, **kwargs):
```
Drive-by cleanup to reduce confusion - This API was already obsolete since 9f1e462ad03708cb7516b64aa08fcaf538028b0a

KevinMind commented on src/olympia/abuse/actions.py:
```diff
@@ -383,10 +383,23 @@ class ContentActionRejectVersionDelayed(ContentActionRejectVersion):
     def __init__(self, decision):
         super().__init__(decision)
-        self.days = int(self.decision.metadata.get('delayed_rejection_days', 0))
+        if 'delayed_rejection_date' in self.decision.metadata:
```
Nit

```python
self.delayed_rejection_date = (
    self.decision.metadata.get('delayed_rejection_date')
    and datetime.fromisoformat(self.decision.metadata['delayed_rejection_date'])
)
```

KevinMind commented on src/olympia/abuse/actions.py:
```diff
@@ -383,10 +383,23 @@ class ContentActionRejectVersionDelayed(ContentActionRejectVersion):
+            self.delayed_rejection_date = datetime.fromisoformat(
+                self.decision.metadata.get('delayed_rejection_date')
+            )
+        else:
```
Is the implicit assumption here that if the value actually is none, then later methods will raise? We should just raise explicitly here then no?

diox replied:
My comment is probably poorly worded. There are times when we need to instantiate the class and call some methods that don't need the `delayed_rejection_date`, and we want those to continue to work.

An example is found in `TestCinderJob.test_report_with_outstanding_rejection` which sends an email to the reporter of an abuse report filed against an add-on with a delayed rejection already set (but that delayed rejection decision doesn't have that metadata - as might be the case for existing ones before this change)

KevinMind replied:
Just curious if that could have been resolved with a migration? All good though no need for a change.

diox replied:
Maybe. We haven't tried to do that in the past, it's a bit of an edge case (we typically don't have to worry about executing past decisions...)

KevinMind commented on src/olympia/abuse/templates/abuse/emails/ContentActionChangePendingRejectionDate.txt:
```diff
@@ -0,0 +1,6 @@
+{% extends "abuse/emails/base.txt" %}{% block content %}
+
+As you are aware, your {{ type }} {{ name }} was manually reviewed by the Mozilla Add-ons team, at which point we found a violation of one or more Mozilla add-on policies.
```
Just my read, but this sounds somewhat passive aggressive.. is the "As you are aware" necessary? I would think this type of email should be more matter of fact and not assume what people may or may not be aware of

diox replied:
In my experience developers are very aware of rejections already :sweat_smile: - and I assume in most cases they are the ones asking for that date to be changed.

Regardless, this is the copy everyone approved (see [issue](https://github.com/mozilla/addons/issues/14920)) so I'm not in hurry to change it and go through validation with Operations, Product and Legal again.

KevinMind commented on src/olympia/abuse/tests/test_models.py:
```diff
@@ -1170,11 +1170,10 @@ def test_process_decision(self):
-        ) as action_mock, mock.patch.object(
-            ContentActionBanUser, 'notify_owners'
-        ) as notify_mock:
+        with (
```
TIL 💪

KevinMind commented on src/olympia/reviewers/forms.py:
```diff
@@ -497,8 +514,54 @@ def clean(self):
                 )
+
+        if self.helper.actions.get(selected_action, {}).get('delayable'):
+            if 'delayed_rejection' not in self.data:
```
It's not totally clear to me how this wouldn't be handled by setting the fields to required (which you do) or using required on the html?

diox replied:
mmm looks like we already manipulate `required` in `is_valid()` above for various fields depending on the action so yeah, I'll try doing that there and drop the custom `ValidationError`

diox replied:
Actually, can't do that because `NullBooleanField` accepts missing values as `None` without raising, even if the field is `required`. I added a comment and stopped hardcoding the error message in 407703a7f58465cf9f570e68eacfdc2534b664d6

KevinMind commented on src/olympia/reviewers/forms.py:
```diff
@@ -497,8 +514,54 @@ def clean(self):
+                    'delayed_rejection',
+                    forms.ValidationError('This field is required.'),
+                )
+            elif self.cleaned_data.get('delayed_rejection') and not self.data.get(
```
Nit: following your other refactorings you are querying the fields from cleaned_data repeatedly and could be extracted to vars..

diox replied:
a1128d64885be43823671b51a83f99ca5e212c7f

KevinMind commented on src/olympia/reviewers/forms.py:
```diff
@@ -497,8 +514,54 @@ def clean(self):
+                if distinct_pending_rejection_dates > 1:
+                    self.add_error(
+                        'versions',
+                        forms.ValidationError(
```
Why is this a requirement? I could imagine having 3 vesrions with different dates and my goal is to set them all to the same single date??

diox replied:
It's a requirement from the issue. I think the main problem is that we have to notify the developer of the old date in the email (see copy you commented on above), which makes it cumbersome if there are multiple different ones. Operations would know more, but @wagnerand noted the main use case would be when dates are all the same anyway so this limitation wouldn't be too bad.

KevinMind replied:
Ok

KevinMind commented on src/olympia/reviewers/tests/test_forms.py:
```diff
@@ -95,7 +95,7 @@ def test_actions_addon_status_null(self):
         assert list(actions.keys()) == [
             'unreject_latest_version',
-            'clear_pending_rejection_multiple_versions',
+            'change_pending_rejection_multiple_versions',
```
nit: it is no longer clear in the name you can also "clear" via this action. Maybe "change_or_clear_pending_rejection_multiple_versions" is super long, but more clear? Or (for some reason) "set" "update" read a bit more broadly to me. Somehow from "change" i'm inferring that there is a value to change to.

diox replied:
e1ffe0983d8264314fe1adbd8353815ef295e3ab

Kept the label the same to avoid wrapping, but updated (eh) the internal name, and the description to mention that changing the date will notify developers.

KevinMind reviewed (commented):
Initial thoughts and feedback/ I'll try to setup and verify tomorrow.

KevinMind commented:
Some feedback after running through the steps

It looks like the change pending rejection is available as an action even if I don't have any pending rejections.. not sure if that is pre-existing but seems odd

<img width="1458" alt="Screenshot 2025-01-31 at 14 04 28" src="https://github.com/user-attachments/assets/24ccea36-0

When I schedule a rejection.. shouldn't I see when the pending rejection is for in the log?

<img width="531" alt="Screenshot 2025-01-31 at 13 27 55" src="https://github.com/user-attachments/assets/a140579b-05d6-4311-8776-fe2051599b8b" />
fc8-4387-91b3-5ea4e1f815b3" />

It is odd that I see an hr/mm at all, and it is actually html required. Can we remove that to only select a date and not time?

<img width="344" alt="Screenshot 2025-01-31 at 13 27 15" src="https://github.com/user-attachments/assets/e0f5f511-ca54-47fd-8b1f-c553e060e165" />

Maybe this is to "react"-ive for what we can do with rev tools but I would expect if I modify the date field, I would exect on blur that we should change the radio to change date and not clear date. It would be super easy to think you are changing the date and actually you clear it. I actually did that before noticing the issue.

https://github.com/user-attachments/assets/e73b2bd3-b6fd-49c8-93a2-184ce850bdfe

diox commented:
> It looks like the change pending rejection is available as an action even if I don't have any pending rejections.. not sure if that is pre-existing but seems odd

That is pre-existing yeah, I didn't change that. We could, I'm not sure it's worth the extra database query (we're trying to minimize them on this page to keep it as fast as possible - every milliseconds counts given how many different things it's loading)

> When I schedule a rejection.. shouldn't I see when the pending rejection is for in the log?

You mean the date ? We don't show it in the log because it's not part of the format string, but we do show the current scheduled rejection date in the row of the version itself in the history if the pending rejection is still active.

> It is odd that I see an hr/mm at all, and it is actually html required. Can we remove that to only select a date and not time?

Maybe. I played with that a little bit, showing the time was a compromise: I was trying to avoid having all scheduled rejections end up at the same time, and not accidentally reduce the delay by a few hours for some developers depending on when the rejection was posted. My original idea was to only allow reviewers to select the date and automatically set the time to the hour/minute of the submission, and in fact that's shown in my screenshots in the description, but then I changed my mind for this implementation to keep things simpler. I don't feel strongly one way or the other though, but I figured, since we do have a full datetime internally, why not show it.

> Maybe this is to "react"-ive for what we can do with rev tools but I would expect if I modify the date field, I would exect on blur that we should change the radio to change date and not clear date. It would be super easy to think you are changing the date and actually you clear it. I actually did that before noticing the issue.

Also something I changed my mind on as I was working on this. My first implementation was more JavaScript-heavy and I went for something a bit simpler to maintain given the current codebase. But I see your point, maybe there is a middle-ground we can reach here. I'll see what I can do.

diox commented:
>> Maybe this is to "react"-ive for what we can do with rev tools but I would expect if I modify the date field, I would exect on blur that we should change the radio to change date and not clear date. It would be super easy to think you are changing the date and actually you clear it. I actually did that before noticing the issue.
>
> Also something I changed my mind on as I was working on this. My first implementation was more JavaScript-heavy and I went for something a bit simpler to maintain given the current codebase. But I see your point, maybe there is a middle-ground we can reach here. I'll see what I can do.

https://github.com/mozilla/addons-server/pull/23001/commits/f1f0ac5243df4fea90a6ad6a5449e683279dc5d1 does this

KevinMind reviewed (approved):
Verified 🚢
